### PR TITLE
[SPARK-8833][STREAMING][WIP] Kafka Direct API support offset in zookeeper

### PR DESCRIPTION
**Pull request by guowei2**
Kafka Direct API support offset in zookeeper

### CODE REVIEW

**Comment by guowei2** [comment]
I will add test case for this later

**Comment by srowen** [comment]
CC @koeninger but I don't think this class is intended to manage Kafka offsets by itself

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 1]
```diff
@@ -402,7 +402,7 @@ object KafkaCluster {
       }
 
       Seq("zookeeper.connect", "group.id").foreach { s =>
-        if (!props.contains(s)) {
+        if (!props.containsKey(s)) {
```
Comment (srowen) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: That's a good bug fix

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 2]
```diff
@@ -403,12 +405,19 @@ object KafkaUtils {
       topicPartitions <- kc.getPartitions(topics).right
       leaderOffsets <- (if (reset == Some("smallest")) {
         kc.getEarliestLeaderOffsets(topicPartitions)
-      } else {
+      } else if (reset == Some("largest")){
         kc.getLatestLeaderOffsets(topicPartitions)
+      } else {
+        val offsets =
+          kc.getConsumerOffsets(kafkaParams.getOrElse("group.id", ""), topicPartitions.seq)
+        if(offsets.isLeft) kc.getLatestLeaderOffsets(topicPartitions)
```
Comment (srowen) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: This style is pretty wrong

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 3]
```diff
@@ -403,12 +405,19 @@ object KafkaUtils {
...
+        else offsets
       }).right
     } yield {
-      val fromOffsets = leaderOffsets.map { case (tp, lo) =>
-          (tp, lo.offset)
+      val fromOffsets = leaderOffsets.map {
+        case (tp, LeaderOffset(_, _, lo)) => (tp, lo)
+        case (tp, lo: Long) => (tp, lo)
+        case _ => throw new SparkException ("get error leader offsets ")
```
Comment (srowen) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: This is not a useful exception

**Comment by koeninger** [comment]
The containsKey thing is a good catch.

But this basic idea has been discussed before and rejected:

https://issues.apache.org/jira/browse/SPARK-6249

I don't think the direct stream should be implicitly doing anything at all with zookeeper.  You can handle writing to / reading from ZK yourself explicitly if you need to.

**Comment by koeninger** [comment]
@guowei2 do you want to open a separate jira / pull request for that containsKey fix?

**Comment by guowei2** [comment]
ok

**Pull request closed without merging**